Karl/README.md

```diff
@@ -0,0 +1,257 @@
+
+**A Directly Responsible Individual who may receive alerts.**
+
+Has a unique name. For the purposes of this doc, assume they have an e-mail address on file.
+  
+  
+#### Service
+
+**A collection of servers to monitor.**
+
+Has a unique name. Each Server to belongs to one and only Service.
+
+#### Server: A URL to connect to a binary that should be up and running
+
+**Effectively, { IP address / DNS name, port, HTTP route }.**
+
+Belongs to one and only Service. Has a unique name within the Service.
+  
+ 
+#### Scrape: A timestamped event of gathering status of a particular server
+
+**Identified by { Timestamp, Service::Server }.**
+
+ID space matches the one of Results.
+
+#### Result: The result of a scrape.
+
+**Identified by { Timestamp, Service::Server }.**
+
+ID space matches the one of Scraped.
+
+#### Alert: An event that is abnormal and should be reported
+
+**Identified by { Service, Name, Timestamps }.**
+
+**Optional { Server, Tags }.**
+
+Name and tags are used for muting, see below.
+
+Most alerts point to specific Scrape-s + Result-s for easier browsing.
+
+#### Page: The alert that has been sent to the DRI
+
+**Identified by { Alert, DRI }.**
+
+Generally, muted alerts are not sent out as pages (yet can be browsed), and everything else makes it to one or more DRI.
+
+#### Mute: A temporary hold placed on an Alert for it to not generate Pages
+
+**Identified by { Service, Alert Name }.**
+
+**Optional: { Alert Tags }.**
+
+#### Dispatcher: The logic that generates pages given alerts and mutes
+
+A single instance per Karl.
+
+By default, muted alerts do not become pages, and non-muted ones are promoted to pages and sent to all DRIs. Paged alerts are then automatically muted by their { Service, Name }, any tags, for the next 15 minutes.
+
+The logic to take into consideration current muting configuration, alert severity, DRI configuration, account medias (e-mail, SMS, etc.), time of day, vacations, etc., is implemented as the Dispatcher.
+
+
+## Implementation
+
+Alerts are defined and configured in a C++ file. No bare strings are used; each entity is defined as a variable of certain type, and then passed over.
+
+A running Karl exposes a dashboard. Karl is also Claire: it is safe to run under `cron`, and it is friendly with continuous integration.
+
+Everything is stored in Yoda, with a REST-ful browsing API for viewing current configuration, scrapes, their results, alerts, pages, and mutes. Mutes can be set, prolonged/shrinked, and unset via this API.
+
+`TODO(dkorolev): Discuss aggregation master-plan with Max.`
+
+For Claire binaries requiring extra time to start up or tear down, pass `karl::WarmupPeriod(FLAGS_warmup_period_ms)` and/or `karl::ShutdownPeriod(FLAGS_shutdown_period_ms)`. This will tell Claire to allow for more time before switching form graceful to non-graceful shutdown of an already running binary in case it is by design expected to take more than the default three seconds to start up or to tear down.
```

Comment:
`in case it is by design expected to take` -> `in case it is expected by design to take`